// @txtextcontrol/tx-ng-document-editor/esm2015/lib/document-editor/document-editor.component.js
import { Component, Input, ElementRef, ViewChild } from '@angular/core';
import { TXSettings } from "./tx-settings";
import { TX_JS } from "./tx-min-js";
import { JQUERY_JS } from "./jquery-js";
import { JQUERY_UI_JS } from "./jquery-ui-js";
import { JQUERY_UI_TOUCH_PUNCH } from "./jquery-ui-touch-punch-js";
import { TX_CSS } from "./tx-css";
import { JQUERY_UI_CSS } from "./jquery-ui-css";
import { Color } from './color';
export class DocumentEditorComponent {
    constructor(hostElement) {
        this.hostElement = hostElement;
        this.documentTargetMarkers = DocumentEditorComponent.DefaultSettings.DocumentTargetMarkers;
        this.backColor = DocumentEditorComponent.DefaultSettings.BackColor.toHexString();
        this.statusBarColor = DocumentEditorComponent.DefaultSettings.StatusBarColor.toHexString();
        this.editMode = 'Edit';
        this.contextMenusEnabled = DocumentEditorComponent.DefaultSettings.ContextMenusEnabled;
        this.saveDocumentQuestionDialogEnabled = DocumentEditorComponent.DefaultSettings.SaveDocumentQuestionDialogEnabled;
        this.tableGridLines = DocumentEditorComponent.DefaultSettings.TableGridLines;
        this.textFrameMarkerLines = DocumentEditorComponent.DefaultSettings.TextFrameMarkerLines;
        this.controlChars = DocumentEditorComponent.DefaultSettings.ControlChars;
        this.textFieldsEditable = DocumentEditorComponent.DefaultSettings.TextFieldsEditable;
        this.formattingPrinter = DocumentEditorComponent.DefaultSettings.FormattingPrinter;
        this.culture = DocumentEditorComponent.DefaultSettings.Culture;
        this.uiCulture = DocumentEditorComponent.DefaultSettings.UICulture;
        this.userNames = DocumentEditorComponent.DefaultSettings.UserNames;
        this.reconnectTimeout = DocumentEditorComponent.DefaultSettings.ReconnectTimeout;
        this.webSocketURL = DocumentEditorComponent.DefaultSettings.WebSocketURL;
        // Generate connection id
        this.connectionID = generateRandomID(21);
    }
    ngOnInit() {
        // Dynamically add necessary scripts and stylesheets to dom
        // jQuery UI CSS
        let elemStyle = document.createElement('style');
        elemStyle.innerText = JQUERY_UI_CSS;
        this.containerDiv.nativeElement.appendChild(elemStyle);
        // TX CSS
        elemStyle = document.createElement('style');
        elemStyle.title = 'TX Text Control HTML5 Editor main Stylesheet';
        elemStyle.innerText = TX_CSS;
        this.containerDiv.nativeElement.appendChild(elemStyle);
        // TX settings
        let elemScript = document.createElement('script');
        elemScript.type = 'application/json';
        elemScript.id = '__txSettings';
        elemScript.innerText = JSON.stringify(this.settings);
        this.containerDiv.nativeElement.appendChild(elemScript);
        // jQuery
        elemScript = document.createElement('script');
        elemScript.innerText = JQUERY_JS;
        this.containerDiv.nativeElement.appendChild(elemScript);
        // jQuery UI
        elemScript = document.createElement('script');
        elemScript.innerText = JQUERY_UI_JS;
        this.containerDiv.nativeElement.appendChild(elemScript);
        // Touch enhancements for jQuery UI
        elemScript = document.createElement('script');
        elemScript.innerText = JQUERY_UI_TOUCH_PUNCH;
        this.containerDiv.nativeElement.appendChild(elemScript);
        // $.noConflict()
        elemScript = document.createElement('script');
        elemScript.innerText = DocumentEditorComponent.NO_CONFLICT_JS;
        this.containerDiv.nativeElement.appendChild(elemScript);
        // TX main script file
        elemScript = document.createElement('script');
        elemScript.textContent = TX_JS;
        this.containerDiv.nativeElement.appendChild(elemScript);
        // Set stuff on host element
        this.hostElement.nativeElement.style.display = 'inline-block';
        this.hostElement.nativeElement.style.width = this.width;
        this.hostElement.nativeElement.style.height = this.height;
        // Fire "editor loaded" event
        var evt = new CustomEvent("txDocumentEditorLoaded", { detail: { /* Empty event argument */} });
        document.dispatchEvent(evt);
    }
    get settings() {
        let settings = new TXSettings(this.webSocketURL);
        settings.ConnectionID = this.connectionID;
        settings.DocumentTargetMarkers = this.documentTargetMarkers;
        settings.BackColor = Color.fromHexString(this.backColor);
        settings.StatusBarColor = Color.fromHexString(this.statusBarColor);
        settings.EditMode = TXSettings.editModeFromString(this.editMode);
        settings.ContextMenusEnabled = this.contextMenusEnabled;
        settings.SaveDocumentQuestionDialogEnabled = this.saveDocumentQuestionDialogEnabled;
        settings.TableGridLines = this.tableGridLines;
        settings.TextFrameMarkerLines = this.textFrameMarkerLines;
        settings.ControlChars = this.controlChars;
        settings.TextFieldsEditable = this.textFieldsEditable;
        settings.FormattingPrinter = this.formattingPrinter;
        settings.Culture = this.culture;
        settings.UICulture = this.uiCulture;
        settings.UserNames = this.userNames;
        settings.ReconnectTimeout = this.reconnectTimeout;
        settings.WebSocketURL = this.webSocketURL;
        return settings;
    }
}
DocumentEditorComponent.DefaultSettings = new TXSettings();
DocumentEditorComponent.NO_CONFLICT_JS = 'var j$1_10_2 = $.noConflict(true);';
DocumentEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'tx-document-editor',
                template: "<div #txTemplateDesignerContainer id=\"txTemplateDesignerContainer\" [style.position]=\"'relative'\">\r\n   <svg id=\"txLoader\" class=\"txLoader\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\"\r\n      style=\"background: rgba(0, 0, 0, 0) none repeat scroll 0% 0%; display: block; shape-rendering: auto;\" width=\"71px\"\r\n      height=\"71px\" viewBox=\"0 0 100 100\" preserveAspectRatio=\"xMidYMid\">\r\n      <circle cx=\"50\" cy=\"50\" r=\"25\" stroke-width=\"5\" stroke=\"#555555\"\r\n         stroke-dasharray=\"39.269908169872416 39.269908169872416\" fill=\"none\" stroke-linecap=\"round\">\r\n         <animateTransform attributeName=\"transform\" type=\"rotate\" dur=\"1.6393442622950818s\" repeatCount=\"indefinite\"\r\n            keyTimes=\"0;1\" values=\"0 50 50;360 50 50\"></animateTransform>\r\n      </circle>\r\n      <circle cx=\"50\" cy=\"50\" r=\"19\" stroke-width=\"5\" stroke=\"#4d83b8\"\r\n         stroke-dasharray=\"29.845130209103033 29.845130209103033\" stroke-dashoffset=\"29.845130209103033\" fill=\"none\"\r\n         stroke-linecap=\"round\">\r\n         <animateTransform attributeName=\"transform\" type=\"rotate\" dur=\"1.6393442622950818s\" repeatCount=\"indefinite\"\r\n            keyTimes=\"0;1\" values=\"0 50 50;-360 50 50\"></animateTransform>\r\n      </circle>\r\n   </svg>\r\n   <div id=\"ribbonbar\" class=\"unselectable\">\r\n      <div id=\"txRibbonCover\"></div>\r\n      <div class=\"txRibbonScrollArrow\" id=\"txRibbonScrollButtonRight\" style=\"opacity: 0\">\r\n         <div class=\"txTriangle\" id=\"txBlackTriangleR\"></div>\r\n      </div>\r\n      <div class=\"txRibbonScrollArrow\" id=\"txRibbonScrollButtonLeft\" style=\"opacity: 0\">\r\n         <div class=\"txTriangle\" id=\"txBlackTriangleL\"></div>\r\n      </div>\r\n      <div id=\"txBtnRibbonHide\">\r\n         <img id=\"txArrowRibbonHideWrapper\"\r\n            src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNgYAAAAAMAASsJTYQAAAAASUVORK5CYII=\" />\r\n      </div>\r\n      <ul class=\"tabs\" style=\"display: none\">\r\n      </ul>\r\n      <div id=\"txRibbonContentArea\">\r\n         <div id=\"txRibbonTabContentContainer\">\r\n         </div>\r\n      </div>\r\n   </div>\r\n   <div id=\"trackChangesSidebar\" class=\"sidebar\" style=\"overflow: hidden;\">\r\n      <img\r\n         src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNgYAAAAAMAASsJTYQAAAAASUVORK5CYII=\"\r\n         class=\"closebtn button\" style=\"width: 16px; height: 16px\" />\r\n      <img\r\n         src=\"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNgYAAAAAMAASsJTYQAAAAASUVORK5CYII=\"\r\n         class=\"refreshbtn button\" style=\"width: 16px; height: 16px\" />\r\n      <div class=\"sidebarHeader unselectable\">\r\n         <span class=\"sidebarHeaderTitle unselectable\"></span>\r\n         <br />\r\n         <span class=\"sidebarHeaderSubTitle unselectable\"></span>\r\n      </div>\r\n      <div class=\"sidebarContentContainer unselectable\">\r\n         <canvas id=\"trackChangesCanvas\" class=\"trackChangesContent\"></canvas>\r\n         <div id=\"txTrackChangesScrollbar\" class=\"txScrollbar txScrollbar-vertical\">\r\n            <div class=\"scroll-button scroll-button-vert scroll-button-up unselectable\">\r\n               <div class=\"scroll-button-arrow-wrapper\"></div>\r\n            </div>\r\n            <div class=\"scrollbar-wrapper unselectable\">\r\n               <div class=\"txslider\"></div>\r\n            </div>\r\n            <div class=\"scroll-button scroll-button-vert scroll-button-down unselectable\"\r\n               style=\"vertical-align: bottom;\">\r\n               <div class=\"scroll-button-arrow-wrapper\"></div>\r\n            </div>\r\n         </div>\r\n         <div id=\"txTrackChangesHorScrollbar\" class=\"txScrollbar txScrollbar-horizontal\">\r\n            <div class=\"scroll-button scroll-button-hor scroll-button-left unselectable\">\r\n               <div class=\"scroll-button-arrow-wrapper\"></div>\r\n            </div>\r\n            <div class=\"scrollbar-wrapper unselectable\">\r\n               <div class=\"txslider\"></div>\r\n            </div>\r\n            <div class=\"scroll-button scroll-button-hor scroll-button-right unselectable\">\r\n               <div class=\"scroll-button-arrow-wrapper\"></div>\r\n            </div>\r\n         </div>\r\n      </div>\r\n   </div>\r\n   <canvas id=\"mainCanvas\" style=\"opacity: 0\"></canvas>\r\n   <canvas id=\"horRulerCanvas\" style=\"opacity: 0\"></canvas>\r\n   <canvas id=\"vertRulerCanvas\" style=\"opacity: 0\"></canvas>\r\n   <div class=\"txScrollbarContainer\" style=\"opacity: 0\">\r\n      <div id=\"scroll-button-left\" class=\"scroll-button scroll-button-hor unselectable\">\r\n         <div class=\"scroll-button-arrow-wrapper\"></div>\r\n      </div>\r\n      <div id=\"horizontal-scrollbar-wrapper\" class=\"scrollbar-wrapper unselectable\">\r\n         <div id=\"slider-horizontal\"></div>\r\n      </div>\r\n      <div id=\"scroll-button-right\" class=\"scroll-button scroll-button-hor unselectable\">\r\n         <div class=\"scroll-button-arrow-wrapper\"></div>\r\n      </div>\r\n   </div>\r\n   <div class=\"txScrollbarContainer\" style=\"opacity: 0\">\r\n      <div id=\"scroll-button-up\" class=\"scroll-button scroll-button-vert unselectable\">\r\n         <div class=\"scroll-button-arrow-wrapper\"></div>\r\n      </div>\r\n      <div id=\"vertical-scrollbar-wrapper\" class=\"scrollbar-wrapper unselectable\">\r\n         <div id=\"slider-vertical\"></div>\r\n      </div>\r\n      <div id=\"scroll-button-down\" class=\"scroll-button scroll-button-vert unselectable\">\r\n         <div class=\"scroll-button-arrow-wrapper\"></div>\r\n      </div>\r\n   </div>\r\n   <canvas id=\"statusBarCanvas\" style=\"opacity: 0\"></canvas>\r\n   <ul id=\"txContextMenu\" class=\"dropDownMenu contextMenu\">\r\n   </ul>\r\n   <form id=\"txNovalidateForm\" novalidate=\"novalidate\" spellcheck=\"false\" autocomplete=\"off\"></form>\r\n</div>",
                styles: [""]
            },] }
];
DocumentEditorComponent.ctorParameters = () => [
    { type: ElementRef }
];
DocumentEditorComponent.propDecorators = {
    containerDiv: [{ type: ViewChild, args: ['txTemplateDesignerContainer', { static: true },] }],
    width: [{ type: Input }],
    height: [{ type: Input }],
    documentTargetMarkers: [{ type: Input }],
    backColor: [{ type: Input }],
    statusBarColor: [{ type: Input }],
    editMode: [{ type: Input }],
    contextMenusEnabled: [{ type: Input }],
    saveDocumentQuestionDialogEnabled: [{ type: Input }],
    tableGridLines: [{ type: Input }],
    textFrameMarkerLines: [{ type: Input }],
    controlChars: [{ type: Input }],
    textFieldsEditable: [{ type: Input }],
    formattingPrinter: [{ type: Input }],
    culture: [{ type: Input }],
    uiCulture: [{ type: Input }],
    userNames: [{ type: Input }],
    reconnectTimeout: [{ type: Input }],
    webSocketURL: [{ type: Input }]
};
const ConIDCharset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890_';
/**
 * Generates a random string with a given length.
 * @param length Length of the generated ID in characters.
 */
function generateRandomID(length) {
    if (length < 1)
        throw new Error('ID must be at least one character long.');
    let result = '';
    for (var i = 0; i < length; ++i) {
        result += ConIDCharset.charAt(Math.floor(Math.random() * ConIDCharset.length));
    }
    return result;
}
