// @txtextcontrol/tx-ng-document-editor/esm2015/lib/document-editor/tx-settings.js
import { Color } from "./color";
export class TXSettings {
    constructor(WebSocketURL = '') {
        this.WebSocketURL = WebSocketURL;
        this.IsMVC = false;
        this.NoPrintEndpoint = true;
        this.NoFileSystemAccess = true; // No file system access on the web server
        this.ConnectionID = '';
        this.DocumentTargetMarkers = true;
        this.BackColor = new Color(0xFF, 0xFF, 0xFF);
        this.StatusBarColor = new Color(0x2B, 0x56, 0x9A);
        this.EditMode = 1;
        this.ContextMenusEnabled = true;
        this.SaveDocumentQuestionDialogEnabled = true;
        this.TableGridLines = true;
        this.TextFrameMarkerLines = true;
        this.ControlChars = false;
        this.TextFieldsEditable = false;
        this.FormattingPrinter = '';
        this.Culture = '';
        this.UICulture = '';
        this.UserNames = [];
        this.ReconnectTimeout = 0;
    }
    static editModeFromString(value) {
        switch (value) {
            case 'Edit': return 1;
            case 'ReadAndSelect': return 2;
            case 'ReadOnly': return 3;
            case 'UsePassword': return 2048;
        }
        return 1; // EditMode.Edit
    }
}
